### ISSUE TITLE
Assertion `brace_count > 0 && "imbalanced brackets are a tokenizer error, we shouldn't be able to get here"' failed.

### ISSUE THREAD

**Issue by faho** [fish-shell/fish-shell#4954]
As reported by @nafg on gitter.

Confirmed with fish 2.7.1-1112-g2fd15e7c

Copy-paste the following:

```fish
curl "https://api.postmarkapp.com/email" \ -X POST \
-H "Accept: application/json" \
-H "Content-Type: application/json" \
-H "X-Postmark-Server-Token: XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX" \
-d "{From: 'sender@example.org', To: 'xxx@gmail.com', Subject: 'Hello from Postmark', HtmlBody: '<strong>Hello</strong> dear Postmark user.'}"
```

navigate to the "sender@example.org" bit and enter some text. That'll trigger the assertion.

This happens because `expand_string()` in the variable stage runs `unescape_string_internal()` with the input set to ` '<strong>Hello</strong> dear Postmark user.'}\"`. That final `}` triggers the assertion, because both the opening `{` and `"` are missed.

`complete_param_expand()` seems to think it's a `--foo:bar` thing because of the `HtmlBody:` - see #2718, where this was also asked for without "--".

I'm not quite sure what to do here - just removing the assert would make this particular case "just work" because there's nothing to complete to begin with, and crashing because something doesn't work in completions seems a bit overzealous.

@ridiculousfish?

**Comment by faho** [comment]
Easier reproducer: Enter this:

```fish
echo "{ : }
```

**Comment by mqudsi** [comment]
I may have been the one to add that assertion when I updated the parser recently. 

The hack employed by the tokenizer to get completions when the text hasn't finished is to fake push the missing tokens to the end. In your test case, there's an open \" which isn't closed, which makes it not match as a string. There's no explicit check for balanced quotes, so they pass on to the brace tokenizer, which fails with that assertion.

I think the "right" thing to do here is to fake-close quotes just like we fake-close braces, brackets, and parentheses.

**Comment by faho** [comment]
>In your test case, there's an open " which isn't closed, which makes it not match as a string.

That's not the case with the longer `curl` example (which has matching quotes and braces), and doesn't quite match what happens, because the assert triggers when it's just the part after the `:`.

>I think the "right" thing to do here is to fake-close quotes just like we fake-close braces, brackets, and parentheses.

The problem isn't the _closing_ bit, but the _opening_ bit that is missed, which makes it think it's not inside double quotes and has no opening brace.

**Comment by mqudsi** [comment]
I'm not sure what I'm missing, as to me it's the same situation in both examples. The key part is that for autocomplete purposes, I don't think the entire command line buffer is tokenized/parsed, only the bit up until the cursor. Normally, the tokenizer moves character by character, processing special characters as they are found. When an unescaped quote is encountered, the entire phrase until the next unescaped quote is treated as a single token and not read into further. 

In both cases, this special behavior that would normally protect against the parsing of content after an unescaped `"` is bypassed (due to the missing/unseen closing `"`) triggering the problem. Unless I'm missing something?

**Comment by faho** [comment]
>for autocomplete purposes, I don't think the entire command line buffer is tokenized/parsed, only the bit up until the cursor

Nope. The relevant bit from `complete.cpp:complete_param_expand()`:

```c++
    if (complete_from_separator) {
        const wcstring sep_string = wcstring(str, sep_index + 1);
        std::vector<completion_t> local_completions;
        if (expand_string(sep_string, &local_completions, flags, NULL) == EXPAND_ERROR) {
.....
    if (complete_from_start) {
        // Don't do fuzzy matching for files if the string begins with a dash (issue #568). We could
        // consider relaxing this if there was a preceding double-dash argument.
        if (string_prefixes_string(L"-", str)) flags &= ~EXPAND_FUZZY_MATCH;

        if (expand_string(str, &this->completions, flags, NULL) == EXPAND_ERROR) {
```

and `str` is the complete current token, as seen in e.g. `complete()`.

The entire token is parsed, but in case of a "separator" (a "=" or ":", for stuff like `--opt=val`) everything _up to_ it is skipped.

>When an unescaped quote is encountered, the entire phrase until the next unescaped quote is treated as a single token and not read into further.

`unescape_string_internal()` (from common.cpp), which is what we're talking about here, _does_ read into stuff further. It gets the single token, and then replaces special characters with our special in-band-signalling private use "characters". E.g. a `~` will be replaced with the `HOME_DIRECTORY` char, while `*` turns into `ANY_STRING`.

We do this as part of expansion - in this particular case, we expand variables by looking for `$` in the appropriate places and turning it in to `VARIABLE_EXPAND` if unquoted and `VARIABLE_EXPAND_SINGLE` if double-quoted.

>this special behavior that would normally protect against the parsing of content after an unescaped " is bypassed

We _need_ the content after `"`. We don't need it after `'` (because variables aren't allowed there).

Also, I just checked and `echo { : bbb }` also works. All that is needed is a `:` or `=` in between braces, quotes aren't necessary.

A real solution would probably entail _first_ unescaping, and then slicing the string, or marking the completion point before unescaping.

However, I'm strongly considering just removing the `assert` for now, because this is not a failure mode that is appropriate for completions.

This was introduced by 24afff1c77c636c1bbc7b1355518313dab9859d1 a bit over a month ago, so it qualifies as a regression.

**Comment by mqudsi** [comment]
Removing the assertion should suffice, though I'd personally comment it out and link here.
